Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java

```diff
@@ -0,0 +1,701 @@
+			sumArea += deltaArea;
+		}
+		return sumArea;
+	}
+
+	static double[] getSurfaceVolume(final Collection<List<Point3f>> surfacePoints) {
+		return surfacePoints.parallelStream().mapToDouble(p -> {
+			if (p == null) {
+				return 0;
+			}
+			final CustomTriangleMesh mesh = new CustomTriangleMesh(p);
+			return Math.abs(mesh.getVolume());
+		}).toArray();
+	}
+
+	/**
+	 * Get the Feret diameter of a surface. Uses an inefficient brute-force
+	 * algorithm.
+	 *
+	 * @param surfacePoints points of all the particles.
+	 * @return Feret diameters of the surfaces.
+	 */
+	static double[] getFerets(final List<List<Point3f>> surfacePoints) {
+		Thread[] threads = Multithreader.newThreads();
+		final int nSurfaces = surfacePoints.size();
+		final double[] ferets = new double[nSurfaces];
+		AtomicInteger ai = new AtomicInteger(0);
+		for (int thread = 0; thread < threads.length; thread++) {
+			threads[thread] = new Thread(() -> {
+				for (int i = ai.getAndIncrement(); i < nSurfaces; i = ai.getAndIncrement()) {
+					final List<Point3f> surface = surfacePoints.get(i);
+				
+					if (surface == null) {
+						ferets[i] = Double.NaN;
+						continue;
+					}
+					
+					// check all the point pairs in this surface
+					ListIterator<Point3f> ita = surface.listIterator();
+					ListIterator<Point3f> itb;
+					Point3f a;
+					Point3f b;
+					double feret = 0;
+					while (ita.hasNext()) {
+						// for all the points
+						a = ita.next();
+						// check all the pairs after this point
+						// (the other direction has already been checked)
+						itb = surface.listIterator(ita.nextIndex());
+						while (itb.hasNext()) {
+							b = itb.next();
+							feret = Math.max(feret, a.distance(b));
+						}
+					}
+					ferets[i] = feret;
+				}
+			});
+		}
+		Multithreader.startAndJoin(threads);
+		return ferets;
+	}
+
+	/**
+	 * Get the list of best-fit ellipsoids for the particle surfaces
+	 * 
+	 * @param surfacePoints
+	 * @return Object[] array containing the list of ellipsoids, each of which is
+	 *         also stored as an Object[] array (see FitEllipsoid.yuryPetrov() for
+	 *         details). Note that an Object[] is also an Object so there is no need
+	 *         to make a 2D array (i.e. Object[][]). However, client code must
+	 *         unwrap the ellipsoid elements into arrays by casting to Object[].
+	 */
+	static Object[] getEllipsoids(final Collection<List<Point3f>> surfacePoints) {
+		return surfacePoints.parallelStream().map(surface -> {
+
+			if (surface == null)
+				return null;
+
+			final Iterator<Point3f> pointIter = surface.iterator();
+			final double[][] coOrdinates = new double[surface.size()][3];
+			int i = 0;
+			while (pointIter.hasNext()) {
+				final Point3f point = pointIter.next();
+				coOrdinates[i][0] = point.x;
+				coOrdinates[i][1] = point.y;
+				coOrdinates[i][2] = point.z;
+				i++;
+			}
+
+			Object[] ellipsoid = null;
+			try {
+				ellipsoid = FitEllipsoid.yuryPetrov(coOrdinates);
```

Comment:
A check and exception should probably go in FitEllipsoid.yuryPetrov() so that calling classes are warned that they are not getting an ellipsoid back for the data they sent to the method.

Reply:
Have a look at [QuadricToEllipsoid](https://github.com/bonej-org/BoneJ2/blob/master/Modern/ops/src/main/java/org/bonej/ops/ellipsoid/QuadricToEllipsoid.java). I found out that you had to do the eigen decomposition for the parameters that come out of YuryPetrov to find out if what you have is indeed an ellipsoid. 

I can't explain exactly why it wasn't enough to look at the signs or determinant of the parameters in the quadric solution, but in practice that caused "this is not an ellipsoid" way too often.

Incidentally, I'm not sure if it makes sense for `QuadricToEllipsoid` to be an op because it is so specific, so if you find that with little modifications you can make it into a utility that benefit both old and modern code, feel free to do so.

Reply:
From my reading it seems sufficient to check that the eigenvalues are all > 0, because the ellipsoid matrix has to be _positive definite_. We have the eigenvalues already to calculate the radii. So just throwing an exception if any of them is <= 0 should be enough:

```
final EigenvalueDecomposition E = (EigenvalueDecomposition) matrices[3];
		final Matrix eVal = E.getD();
		final Matrix diagonal = MatrixUtils.diag(eVal);
		final int nEvals = diagonal.getRowDimension();
		final double[] radii = new double[nEvals];
		for (int i = 0; i < nEvals; i++) {
			final double eigenvalue = diagonal.get(i, 0);
			if (eigenvalue <= 0) {
				throw new IllegalArgumentException("Ellipsoid eigenvalues must be greater than 0 "
						+ "(positive definite matrix)");
			}
			radii[i] = Math.sqrt(1 / eigenvalue);
		}
```
Incidentally, this check prevents a NaN result or Math.sqrt spewing because it prevents calculation of 1/0 and Math.sqrt(negative number).